File: frontend/src/lib/projects.ts
import { apiFetch } from "@/lib/api";

export type ProjectStatus = "active" | "paused" | "done";

export type Project = {
  id: number;
  name: string;
  objective: string;
  description: string;
  status: ProjectStatus;
  deadline: string | null;
  created_by: number;
  created_at: string;
  updated_at: string;
};

export type ProjectCreate = {
  name: string;
  objective: string;
  description: string;
  deadline?: string | null;
};

export type ProjectUpdate = {
  name?: string;
  objective?: string;
  description?: string;
  status?: ProjectStatus;
  deadline?: string | null;
};

export async function getProjects(): Promise<Project[]> {
  return apiFetch("/projects");
}

export async function getProject(
  id: number,
): Promise<Project> {
  return apiFetch(`/projects/${id}`);
}

export async function createProject(
  payload: ProjectCreate,
): Promise<Project> {
  return apiFetch("/projects", {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export async function updateProject(
  id: number,
  payload: ProjectUpdate,
): Promise<Project> {
  return apiFetch(`/projects/${id}`, {
    method: "PATCH",
    body: JSON.stringify(payload),
  });
}

export async function deleteProject(
  id: number,
): Promise<void> {
  await apiFetch(`/projects/${id}`, {
    method: "DELETE",
  });
}

File: backend/app/api/projects.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy import select
from sqlalchemy.orm import Session

from app.api.dependencies import (
    get_current_user,
    require_roles,
)
from app.core.enums import UserRole
from app.db.database import get_db
from app.models.project import Project
from app.models.user import User
from app.schemas.project import (
    ProjectCreate,
    ProjectUpdate,
    ProjectResponse,
)


router = APIRouter(
    prefix="/projects",
    tags=["Projects"],
)


@router.post(
    "",
    response_model=ProjectResponse,
    status_code=status.HTTP_201_CREATED,
)
def create_project(
    payload: ProjectCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(
        require_roles(
            UserRole.ADMIN,
            UserRole.HEAD,
        )
    ),
):
    project = Project(
        name=payload.name,
        objective=payload.objective,
        description=payload.description,
        deadline=payload.deadline,
        created_by=current_user.id,
    )

    db.add(project)
    db.commit()
    db.refresh(project)

    return project


@router.get(
    "",
    response_model=list[ProjectResponse],
)
def get_projects(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    projects = db.scalars(
        select(Project).order_by(
            Project.created_at.desc()
        )
    ).all()

    return projects


@router.get(
    "/{project_id}",
    response_model=ProjectResponse,
)
def get_project(
    project_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    project = db.get(
        Project,
        project_id,
    )

    if not project:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Project not found",
        )

    return project


@router.patch(
    "/{project_id}",
    response_model=ProjectResponse,
)
def update_project(
    project_id: int,
    payload: ProjectUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(
        require_roles(
            UserRole.ADMIN,
            UserRole.HEAD,
        )
    ),
):
    project = db.get(
        Project,
        project_id,
    )

    if not project:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Project not found",
        )

    updates = payload.model_dump(
        exclude_unset=True
    )

    for field, value in updates.items():
        setattr(
            project,
            field,
            value,
        )

    db.commit()
    db.refresh(project)

    return project


@router.delete(
    "/{project_id}",
    status_code=status.HTTP_204_NO_CONTENT,
)
def delete_project(
    project_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(
        require_roles(UserRole.ADMIN)
    ),
):
    project = db.get(
        Project,
        project_id,
    )

    if not project:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Project not found",
        )

    db.delete(project)
    db.commit()